### PR TITLE
Fix missing ExampleWeightColumnName in the advanced Options for some trainers

### LINKED ISSUE #2175: ML.NET public API  exposes parameter `weights` which is not used.
For some of the learners e.g. SdcaRegression  ML.NET public API  exposes parameter `weights`

https://github.com/dotnet/machinelearning/blob/cabf55b543129e8962ba716e8e22c72ac08773e3/src/Microsoft.ML.StandardLearners/StandardLearnersCatalog.cs#L65-L70

However, the advanced `Options` for  `SdcaRegressionTrainer` do not have a field for 'WeightColumn'.  Also, I believe  the algo itself does not use  weights  (need to verify)

We need to scrub our public API  for such spurious uses of  `weights` parameter.

EDIT :  
- Based on investigations (see below), we see that the `SdcaRegressionTrainer`  does use the weights column.  So the proper fix would be to ensure that the weight column passed in the constructor gets used by the trainer. 
- For the example above, we need to add `WeightColumn` in the advanced `Options` for  `SdcaRegressionTrainer`  and verify it gets used by the SDCA trainer.

@sfilipi @glebuk @TomFinley

### PR DESCRIPTION
**Pull request by abgoswam**
Fixes #2175 

- Added `ExampleWeightColumnName` in the advanced `Options` for SDCA trainers {Regression, LogisticRegression, MaximumEntropy}
- Added tests

### CODE REVIEW

File: src/Microsoft.ML.StandardTrainers/Standard/SdcaBinary.cs [hunk 1]
```diff
@@ -154,7 +154,7 @@ public abstract class SdcaTrainerBase<TOptions, TTransformer, TModel> : Stochast
         /// <summary>
         /// Options for the SDCA-based trainers.
         /// </summary>
-        public abstract class OptionsBase : TrainerInputBaseWithLabel
+        public abstract class OptionsBase : TrainerInputBaseWithWeight
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.StandardTrainers/Standard/SdcaBinary.cs]: >TrainerInputBaseWithWeight [](start = 44, length = 26)

I would assume you check all mlContext.*Catalog extensions for SDCA trainers to have exampleWeightColumnName in it?

Comment (abgoswam) [reply src/Microsoft.ML.StandardTrainers/Standard/SdcaBinary.cs]: yeap. i did verify all of the  `simple`  SDCA trainer extensions  have the `exampleWeightColumnName` parameter

---
In reply to: [269666016](https://github.com/dotnet/machinelearning/pull/3104#discussion_r269666016) [](ancestors = 269666016)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs [hunk 2]
```diff
@@ -12,7 +12,7 @@ public static void Example()
             var mlContext = new MLContext();
 
             // Get a small dataset as an IEnumerable and them read it as ML.NET's data type.
-            IEnumerable<SamplesUtils.DatasetUtils.BinaryLabelFloatFeatureVectorSample> enumerableOfData = SamplesUtils.DatasetUtils.GenerateBinaryLabelFloatFeatureVectorSamples(5);
+            IEnumerable<SamplesUtils.DatasetUtils.BinaryLabelFloatFeatureVectorFloatWeightSample> enumerableOfData = SamplesUtils.DatasetUtils.GenerateBinaryLabelFloatFeatureVectorFloatWeightSamples(5);
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: How do you examine the effect of having weight? Should it be somehow checked in a test?

Comment (abgoswam) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: Thats what i did in the test `SdcaLogisticRegressionWithWeight` .. Added two  trainers w/wo weights and verified it produced different metrics.  Is that sufficient ?

---
In reply to: [270204851](https://github.com/dotnet/machinelearning/pull/3104#discussion_r270204851) [](ancestors = 270204851)

File: test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs [hunk 3]
```diff
@@ -88,11 +88,66 @@ public void SdcaLogisticRegression()
...
+            // algorithms which needs many data passes. Since SDCA is the case, we cache.
+            data = mlContext.Data.Cache(data);
+
+            // Verify SdcaLogisticRegression with and without weights.
+            var sdcaWithoutWeightBinary = mlContext.BinaryClassification.Trainers.SdcaLogisticRegression(
+                new SdcaLogisticRegressionBinaryTrainer.Options { NumberOfThreads = 1 });
+            var sdcaWithWeightBinary = mlContext.BinaryClassification.Trainers.SdcaLogisticRegression(
+                new SdcaLogisticRegressionBinaryTrainer.Options { ExampleWeightColumnName = "Weight", NumberOfThreads = 1 });
+
+            var prediction1 = sdcaWithoutWeightBinary.Fit(data).Transform(data);
```
Comment (wschin) [inline test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: Could you check that if all (or most) results in `prediction1` and `prediction2` are different?

File: test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs [hunk 4]
```diff
@@ -88,11 +88,66 @@ public void SdcaLogisticRegression()
...
+            var metrics1 = mlContext.BinaryClassification.Evaluate(prediction1);
+            var metrics2 = mlContext.BinaryClassification.Evaluate(prediction2);
+
+            Assert.Equal(0.9658, metrics1.AreaUnderRocCurve, 4);
+            Assert.Equal(0.3488, metrics1.LogLoss, 4);
+            Assert.Equal(0.9596, metrics2.AreaUnderRocCurve, 4);
+            Assert.Equal(0.3591, metrics2.LogLoss, 4);
+
+            // Verify SdcaMaximumEntropy with and without weights.
+            var sdcaWithoutWeightMulticlass = mlContext.Transforms.Conversion.MapValueToKey("LabelIndex", "Label").
```
Comment (wschin) [inline test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: May we put multi-class test to another independent test (to make test small)?

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs (hunk 2 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: Their scores are similar. I'd like to have a more strict criterion. As you heard from Zeeshan S, tiny changes induced large SDCA regression this morning.

---
In reply to: [270205745](https://github.com/dotnet/machinelearning/pull/3104#discussion_r270205745) [](ancestors = 270205745,270204851)

File: test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs [hunk 5]
```diff
@@ -88,11 +88,66 @@ public void SdcaLogisticRegression()
             Assert.InRange(first.Probability, 0.8, 1);
         }
 
+        [Fact]
+        public void SdcaLogisticRegressionWithWeight()
```
Comment (wschin) [inline test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: >LogisticRegression [](start = 24, length = 18)

This is called `LogisticRegression` but contains `MaximumEntropy` trainers.

Comment (abgoswam) [reply test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: WIll fix by separating into 2 tests one each for  binary and multiclass 

---
In reply to: [270208235](https://github.com/dotnet/machinelearning/pull/3104#discussion_r270208235) [](ancestors = 270208235)

File: test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs (hunk 3 above)
Comment (abgoswam) [reply test/Microsoft.ML.Tests/TrainerEstimators/SdcaTests.cs]: added checks to ensure the model parameters for both models are different

---
In reply to: [270206386](https://github.com/dotnet/machinelearning/pull/3104#discussion_r270206386) [](ancestors = 270206386)

File: docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs (hunk 2 above)
Comment (abgoswam) [reply docs/samples/Microsoft.ML.Samples/Dynamic/DataOperations/BootstrapSample.cs]: added checks in the other tests where we test thoroughly for this

---
In reply to: [270207991](https://github.com/dotnet/machinelearning/pull/3104#discussion_r270207991) [](ancestors = 270207991,270205745,270204851)

**Pull request merged**